File: app/src/main/java/sfllhkhan95/doodle/core/views/OptionsDialog.java
package sfllhkhan95.doodle.core.views;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdView;

import sfllhkhan95.doodle.R;
import sfllhkhan95.doodle.ads.AdManager;

/**
 * @author saifkhichi96
 * @version 1.1.0
 * @since 3.3.0
 */
public class OptionsDialog extends Dialog {

    private ImageView dialogIcon;
    private TextView titleView;
    private TextView messageView;

    private String option1;
    private ImageView option1Icon;
    private TextView option1Label;
    private View.OnClickListener op1Listener;

    private String option2;
    private ImageView option2Icon;
    private TextView option2Label;
    private View.OnClickListener op2Listener;

    private String option3;
    private ImageView option3Icon;
    private TextView option3Label;
    private View.OnClickListener op3Listener;

    private String title;
    private String message;

    @DrawableRes
    private int iconId = -1;

    @DrawableRes
    private int iconOp1Id = -1;

    @DrawableRes
    private int iconOp2Id = -1;

    @DrawableRes
    private int iconOp3Id = -1;

    private boolean labelsEnabled = false;

    private boolean adsDisabled = true;

    private OptionsDialog(@NonNull Context context) {
        super(context);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getWindow() != null) {
            getWindow().setBackgroundDrawableResource(android.R.color.transparent);
        }
        setContentView(R.layout.dialog_options);

        // Get UI elements
        dialogIcon = findViewById(R.id.icon);
        titleView = findViewById(R.id.title);
        messageView = findViewById(R.id.message);

        option1Icon = findViewById(R.id.op1_icon);
        option1Label = findViewById(R.id.op1_label);

        option2Icon = findViewById(R.id.op2_icon);
        option2Label = findViewById(R.id.op2_label);

        option3Icon = findViewById(R.id.op3_icon);
        option3Label = findViewById(R.id.op3_label);

        // Configure UI elements
        if (iconId != -1) {
            dialogIcon.setImageResource(iconId);
        }
        titleView.setText(title);
        messageView.setText(message);

        findViewById(R.id.op1).setVisibility(View.GONE);
        findViewById(R.id.op2).setVisibility(View.GONE);
        findViewById(R.id.op3).setVisibility(View.GONE);

        if (iconOp1Id != -1) {
            findViewById(R.id.op1).setVisibility(View.VISIBLE);
            option1Icon.setImageResource(iconOp1Id);
            option1Label.setText(option1);

            option1Icon.setOnClickListener(op1Listener);
            option1Label.setOnClickListener(op1Listener);
        }

        if (iconOp2Id != -1) {
            findViewById(R.id.op2).setVisibility(View.VISIBLE);
            option2Icon.setImageResource(iconOp2Id);
            option2Label.setText(option2);

            option2Icon.setOnClickListener(op2Listener);
            option2Label.setOnClickListener(op2Listener);
        }

        if (iconOp3Id != -1) {
            findViewById(R.id.op3).setVisibility(View.VISIBLE);
            option3Icon.setImageResource(iconOp3Id);
            option3Label.setText(option3);

            option3Icon.setOnClickListener(op3Listener);
            option3Label.setOnClickListener(op3Listener);
        }

        if (labelsEnabled) {
            showLabels();
        } else {
            hideLabels();
        }

        // Display ads if they are enabled
        if (!adsDisabled) {
            final AdView mAdView = this.findViewById(R.id.adView);
            AdManager.loadBanner(mAdView, new AdListener() {
                @Override
                public void onAdLoaded() {
                    super.onAdLoaded();
                    mAdView.setVisibility(View.VISIBLE);
                }
            });
        }
    }

    public void showLabels() {
        option1Label.setVisibility(View.VISIBLE);
        option2Label.setVisibility(View.VISIBLE);
        option3Label.setVisibility(View.VISIBLE);
    }

    private void hideLabels() {
        option1Label.setVisibility(View.GONE);
        option2Label.setVisibility(View.GONE);
        option3Label.setVisibility(View.GONE);
    }

    public static class Builder {

        private final OptionsDialog optionsDialog;

        public Builder(Context context) {
            optionsDialog = new OptionsDialog(context);
        }

        public Builder setIcon(int iconId) {
            optionsDialog.iconId = iconId;
            return this;
        }

        public Builder setTitle(String title) {
            optionsDialog.title = title;
            return this;
        }

        public Builder setMessage(String message) {
            optionsDialog.message = message;
            return this;
        }

        public Builder setOption1(String label, @DrawableRes int iconRes, View.OnClickListener clickListener) {
            optionsDialog.option1 = label;
            optionsDialog.iconOp1Id = iconRes;
            optionsDialog.op1Listener = clickListener;
            return this;
        }

        public Builder setOption2(String label, @DrawableRes int iconRes, View.OnClickListener clickListener) {
            optionsDialog.option2 = label;
            optionsDialog.iconOp2Id = iconRes;
            optionsDialog.op2Listener = clickListener;
            return this;
        }

        public Builder setOption3(String label, @DrawableRes int iconRes, View.OnClickListener clickListener) {
            optionsDialog.option3 = label;
            optionsDialog.iconOp3Id = iconRes;
            optionsDialog.op3Listener = clickListener;
            return this;
        }

        public OptionsDialog create() {
            return optionsDialog;
        }

        public Builder setLabelsEnabled(boolean labelsEnabled) {
            optionsDialog.labelsEnabled = labelsEnabled;
            return this;
        }

        public Builder setAdsDisabled(boolean adsDisabled) {
            optionsDialog.adsDisabled = adsDisabled;
            return this;
        }
    }
}

File: app/src/main/java/sfllhkhan95/doodle/views/dialog/OptionsDialog.kt
package sfllhkhan95.doodle.views.dialog

import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdView
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import sfllhkhan95.doodle.R
import sfllhkhan95.doodle.bo.AdManager
import sfllhkhan95.doodle.utils.ThemeUtils

/**
 * @author saifkhichi96
 * @version 1.1.0
 * @since 3.3.0
 */
class OptionsDialog private constructor() : BottomSheetDialogFragment() {

    private val options: MutableList<Option> = mutableListOf()
    private var title: String? = null
    private var message: String? = null

    private class Option {
        @DrawableRes
        var icon = -1

        lateinit var title: String
        lateinit var description: String
        lateinit var clickListener: DialogInterface.OnClickListener
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(DialogFragment.STYLE_NO_FRAME, ThemeUtils.getDialogTheme())
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.dialog_options, container, false)
        dialog?.window?.setBackgroundDrawableResource(android.R.color.transparent)
        dialog?.window?.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)

        // Inflate options
        v?.findViewById<LinearLayout>(R.id.optionsList)?.let { optionsList ->
            for (option in options) {
                View.inflate(context, R.layout.view_dialog_option, optionsList)
                val v1 = optionsList.getChildAt(optionsList.childCount - 1)

                v1.findViewById<TextView>(R.id.optionText)?.text = option.title
                v1.findViewById<TextView>(R.id.optionDescription)?.text = option.description
                if (option.icon != -1) {
                    v1.findViewById<ImageView>(R.id.optionIcon)?.setImageResource(option.icon)
                }
                v1.setOnClickListener {
                    option.clickListener.onClick(this@OptionsDialog.dialog, 0)
                }
            }
        }

        // Display ads if they are enabled
        v?.findViewById<AdView>(R.id.adView)?.let {
            AdManager.instance.showBannerAd(it, object : AdListener() {
                override fun onAdLoaded() {
                    super.onAdLoaded()
                    it.visibility = View.VISIBLE
                }
            })
        }

        title?.let {
            v.findViewById<TextView>(R.id.title).text = it
            v.findViewById<TextView>(R.id.title).visibility = View.VISIBLE
        }

        message?.let {
            v.findViewById<TextView>(R.id.message).text = it
            v.findViewById<TextView>(R.id.message).visibility = View.VISIBLE
        }

        return v
    }

    fun show(manager: FragmentManager) {
        super.show(manager, "options_dialog")
    }

    class Builder {

        private val optionsDialog = OptionsDialog()

        fun addOption(label: String, description: String, @DrawableRes iconRes: Int, clickListener: DialogInterface.OnClickListener): Builder {
            optionsDialog.options.add(Option().apply {
                this.icon = iconRes
                this.title = label
                this.description = description
                this.clickListener = clickListener
            })
            return this
        }

        fun setTitle(title: String): Builder {
            optionsDialog.title = title
            return this
        }

        fun setMessage(message: String): Builder {
            optionsDialog.message = message
            return this
        }

        fun create(): OptionsDialog {
            return optionsDialog
        }

    }

}